### PR TITLE
Add catalyst provider support

### PR DESCRIPTION
**Pull request by tkstanczak**
### New Pull Request Submissions:

1. [x] Have you followed the guidelines in our [Contributing document](https://github.com/atlascity/Community/tree/master/CONTRIBUTING.md)?
2. [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
3. [x] I have added tests to cover my changes.
4. [x] All new and existing tests passed.
5. [ ] Have you lint your code locally prior to submission?
6. [x] Does your code follows the code style of this project?
7. [ ] Does your change require a change to the documentation.
    - [ ] I have updated the documentation accordingly.
9. [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
10. [ ] Have you inserted a keyword and link to the issues the PR closes in its descriptions (ex `closes #1`) ?
11. [ ] Is you branch up to date, have you integrated all the latest changes from develop and resolved conflicts ?

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**
Adds Catalyst Wallet Provider support within Web3
Adds batch requests handling

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

**Comment by nshCore** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Clones added
============
- src/Catalyst.Core.Modules.Ledger/Web3EthApi.cs  1
- src/Catalyst.Core.Lib/IO/Events/TransactionReceivedEvent.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051857&bid=16632740)

File: src/Catalyst.Core.Modules.Ledger/Web3EthApi.cs [hunk 1]
```diff
@@ -88,7 +91,17 @@ public Keccak SendTransaction(PublicEntry publicEntry)
                 PublicEntry = publicEntry
             };
 
+            publicEntry.Timestamp = Timestamp.FromDateTime(DateTime.UtcNow);
             _transactionReceived.OnTransactionReceived(broadcast.ToProtocolMessage(_peerId));
+            
+            byte[] kvmAddressBytes = Keccak.Compute(publicEntry.SenderAddress.ToByteArray()).Bytes.AsSpan(12).ToArray();
+            string hex = kvmAddressBytes.ToHexString() ?? throw new ArgumentNullException("kvmAddressBytes.ToHexString()");
```
Comment (nshCore) [inline src/Catalyst.Core.Modules.Ledger/Web3EthApi.cs]: Codacy Issue found: [The parameter name 'kvmAddressBytes.ToHexString()' is not declared in the argument list.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051857)

File: src/Catalyst.Core.Lib.Tests/UnitTests/IO/Events/TransactionReceivedEventTests.cs [hunk 2]
```diff
@@ -66,7 +68,7 @@ public void Can_Send_Error_To_Invalid_Transaction()
         {
             _transactionValidator.ValidateTransaction(Arg.Any<PublicEntry>())
                .Returns(false);
-            _transactionReceivedEvent.OnTransactionReceived(new TransactionBroadcast {PublicEntry = new PublicEntry()}
+            _transactionReceivedEvent.OnTransactionReceived(new TransactionBroadcast {PublicEntry = new PublicEntry{SenderAddress = new byte[32].ToByteString(), Timestamp = Timestamp.FromDateTime(DateTime.UtcNow)}}
```
Comment (nshCore) [inline src/Catalyst.Core.Lib.Tests/UnitTests/IO/Events/TransactionReceivedEventTests.cs]: Codacy Issue found: [Split this 214 characters long line (which is greater than 200 authorized).](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051857)

File: src/Catalyst.Core.Modules.Consensus/Deltas/DeltaCache.cs [hunk 3]
```diff
@@ -55,6 +55,8 @@ public class DeltaCache : IDeltaCache, IDisposable
 
         public static Address TruffleTestAccount = new Address("0xb77aec9f59f9d6f39793289a09aea871932619ed");
 
+        public static Address CatalystTruffleTestAccount = new Address("0x58BeB247771F0B6f87AA099af479aF767CcC0F00");
```
Comment (nshCore) [inline src/Catalyst.Core.Modules.Consensus/Deltas/DeltaCache.cs]: Codacy Issue found: [Change the visibility of 'CatalystTruffleTestAccount' or make it 'const' or 'readonly'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051857)

File: src/Catalyst.Core.Lib/IO/Events/TransactionReceivedEvent.cs [hunk 4]
```diff
@@ -59,10 +64,23 @@ public class TransactionReceivedEvent : ITransactionReceivedEvent
...
             {
-                return ResponseCode.Error;
+                var transactionValid = _validator.ValidateTransaction(publicEntry);
+                if (!transactionValid)
+                {
+                    return ResponseCode.Error;
+                }
+                
+                byte[] kvmAddressBytes = Keccak.Compute(publicEntry.SenderAddress.ToByteArray()).Bytes.AsSpan(12).ToArray();
+                string hex = kvmAddressBytes.ToHexString() ?? throw new ArgumentNullException("kvmAddressBytes.ToHexString()");
```
Comment (nshCore) [inline src/Catalyst.Core.Lib/IO/Events/TransactionReceivedEvent.cs]: Codacy Issue found: [Remove the unused local variable 'hex'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051857)

File: src/Catalyst.Core.Modules.Consensus/Deltas/DeltaCache.cs (hunk 3 above)
Comment (nshCore) [inline src/Catalyst.Core.Modules.Consensus/Deltas/DeltaCache.cs]: Codacy Issue found: [Make this field 'private' and encapsulate it in a 'public' property.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051857)

File: src/Catalyst.Core.Modules.Web3/Controllers/EthController.cs [hunk 5]
```diff
@@ -47,9 +49,33 @@ public EthController(IWeb3EthApi web3EthApi, IWeb3HandlerResolver handlerResolve
...
+        }
 
         [HttpPost]
-        public JsonRpcResponse Request([FromBody] JsonRpcRequest request)
+        public object Request([FromBody] object body)
+        {
+            JObject jObject = body as JObject;
+            if (jObject?.Type == JTokenType.Array)
+            {
+                JsonRpcRequest[] requests = jObject.ToObject<JsonRpcRequest[]>();
```
Comment (nshCore) [inline src/Catalyst.Core.Modules.Web3/Controllers/EthController.cs]: Codacy Issue found: ['jObject' is null on at least one execution path.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051857)

File: src/Catalyst.Core.Modules.Ledger/Web3EthApi.cs (hunk 1 above)
Comment (nshCore) [inline src/Catalyst.Core.Modules.Ledger/Web3EthApi.cs]: Codacy Issue found: [Remove the unused local variable 'hex'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051857)

File: src/Catalyst.Core.Lib/IO/Events/TransactionReceivedEvent.cs (hunk 4 above)
Comment (nshCore) [inline src/Catalyst.Core.Lib/IO/Events/TransactionReceivedEvent.cs]: Codacy Issue found: [The parameter name 'kvmAddressBytes.ToHexString()' is not declared in the argument list.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051857)

File: src/Catalyst.Core.Modules.Ledger.Tests/IntegrationTests/TruffleTest/package.json [hunk 6]
```diff
@@ -20,6 +20,7 @@
   },
   "homepage": "https://github.com/vicnaum/truffleTest#readme",
   "dependencies": {
+    "@catalyst-net-js/truffle-provider": "0.0.9",
```
Comment (nshCore) [inline src/Catalyst.Core.Modules.Ledger.Tests/IntegrationTests/TruffleTest/package.json]: This js library is published under the wrong npm organisation @stxphxn remove this org please and publish under @catalystnetwork

File: src/Catalyst.Core.Modules.Ledger.Tests/IntegrationTests/TruffleTest/truffle-config.js [hunk 7]
```diff
@@ -18,7 +18,8 @@
  *
  */
 
-const HDWalletProvider = require('truffle-hdwallet-provider');
+const {HDWalletProvider} = require("@catalyst-net-js/truffle-provider");
```
Comment (nshCore) [inline src/Catalyst.Core.Modules.Ledger.Tests/IntegrationTests/TruffleTest/truffle-config.js]: Same comment as above @stxphxn publish under the correct org

File: src/Catalyst.Core.Modules.Web3/Controllers/EthController.cs [hunk 8]
```diff
@@ -47,9 +49,33 @@ public EthController(IWeb3EthApi web3EthApi, IWeb3HandlerResolver handlerResolve
...
+            }
+
+            return responses;
+        }
 
         [HttpPost]
-        public JsonRpcResponse Request([FromBody] JsonRpcRequest request)
+        public object Request([FromBody] object body)
+        {
+            JObject jObject = body as JObject;
```
Comment (stxphxn) [inline src/Catalyst.Core.Modules.Web3/Controllers/EthController.cs]: jObject returns null if body is an array. Should it be JToken?

Comment (tkstanczak) [reply src/Catalyst.Core.Modules.Web3/Controllers/EthController.cs]: fixed, thanks, we need to add tests for this

**Pull request merged**